components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIKeyValidator.java

```diff
@@ -701,15 +708,53 @@ public VerbInfoDTO getVerbInfoDTOFromAPIData(String context, String apiVersion,
 
 
     @MethodStats
-    protected ArrayList<URITemplate> getAllURITemplates(String context, String apiVersion)
+    protected ArrayList<URITemplate> getAllURITemplates(MessageContext messageContext, String context, String apiVersion)
             throws APISecurityException {
-        return dataStore.getAllURITemplates(context, apiVersion);
+        if (uriTemplates == null) {
+            synchronized (this) {
+                if (uriTemplates == null) {
```

Comment:
Does this condition is required to re-check? what about L 713?

Reply:
@HiranyaKavishani - I think is needed due to the synchronized block. This is called the double check.